### PR TITLE
Experiment.astazero

### PR DESCRIPTION
**Pull request by bjornborg**

### CODE REVIEW

File: code/system/application/knowledge/include/linguistics/v2vcam/v2vcam.hpp [hunk 1]
```diff
@@ -48,24 +51,51 @@ class V2vCam
...
+  int32_t GetVehicleWidth();
+  int32_t GetLongitudinalAcc();
+  int32_t GetLongitudinalAccConf();
+  int32_t GetYawRateValue();
+  int32_t GetYawRateConfidence();
+  int32_t GetVehicleRole();
+
   std::ofstream m_sendLog;
   std::ofstream m_receiveLog;
+  std::time_t m_tt;
```
Comment (olbender) [inline code/system/application/knowledge/include/linguistics/v2vcam/v2vcam.hpp]: This name can be improved.

File: code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
+#ifndef __INVERSEPERSPECTIVEMAPPING_HPP_
+#define __INVERSEPERSPECTIVEMAPPING_HPP_
```
Comment (olbender) [inline code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp]: Both lines: The keyword should be DETECTLANE_INVERSEPERSPECTIVEMAPPING_HPP_

File: code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+#include <iostream>
+
+class InversePerspectiveMapping
+{
+
+public:
+	InversePerspectiveMapping();
+	InversePerspectiveMapping(const cv::Size &a_originalSize,const cv::Size &a_outputSize,
+			const std::vector<cv::Point2f>& a_regionPoints, const std::vector<cv::Point2f>& a_outputPoints);
+	void applyHomography(cv::Mat &a_inputImage,cv::Mat &a_outputImage, int borderMode = cv::BORDER_CONSTANT);
```
Comment (olbender) [inline code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp]: Methods should start with capital letter.

File: code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+#include <Eigen/Dense>
+#include <iostream>
+
+class InversePerspectiveMapping
+{
+
+public:
+	InversePerspectiveMapping();
+	InversePerspectiveMapping(const cv::Size &a_originalSize,const cv::Size &a_outputSize,
```
Comment (olbender) [inline code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp]: 'const' should be after type (usually the type is of more interest).

File: code/system/application/perception/include/detectlane/detectlane.hpp [hunk 5]
```diff
@@ -42,8 +45,64 @@ class DetectLane
   virtual void nextContainer(odcore::data::Container &);
 
  private:
+  bool m_setup;
```
Comment (olbender) [inline code/system/application/perception/include/detectlane/detectlane.hpp:57]: Methods should always be listed first.

File: code/system/application/perception/src/detectlane/InversePerspectiveMapping.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+void InversePerspectiveMapping::applyHomography(cv::Mat &a_inputImage,cv::Mat &a_outputImage, int a_borderCondition)
+{
+	cv::remap(a_inputImage,a_outputImage,m_gridX,m_gridY,cv::INTER_LINEAR,a_borderCondition);
+	
+}
+
+void InversePerspectiveMapping::drawPoints( const std::vector<cv::Point2f>& _points, cv::Mat& _img )
+{
+	cv::line(_img, cv::Point(static_cast<int>(_points[0].x), static_cast<int>(_points[0].y)), cv::Point(static_cast<int>(_points[3].x), static_cast<int>(_points[3].y)), CV_RGB( 205,205,0), 2);
```
Comment (olbender) [inline code/system/application/perception/src/detectlane/InversePerspectiveMapping.cpp]: Break lines which are longer than 80 characters.

File: code/system/application/perception/src/detectlane/InversePerspectiveMapping.cpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+
+void InversePerspectiveMapping::generateRemappingFunction()
+{
+	// Create remap images
+	m_gridX.create(m_outputSize, CV_32F);
+	m_gridY.create(m_outputSize, CV_32F);
+	for( int j = 0; j < m_outputSize.height; ++j )
+	{
+		float* rowPointX = m_gridX.ptr<float>(j);
```
Comment (olbender) [inline code/system/application/perception/src/detectlane/InversePerspectiveMapping.cpp]: Use indent of 2 spaces

File: code/system/application/perception/src/detectlane/detectlane.cpp [hunk 8]
```diff
@@ -70,266 +101,359 @@ void DetectLane::nextContainer(odcore::data::Container &c)
...
-    //std::cout<<"Offset "<<laneOffset<<std::endl;
-		
-    
-    // Send the message
-    opendlv::perception::LanePosition lanePosition(laneOffset,theta);
-    odcore::data::Container msg(lanePosition);  
-    getConference().send(msg);
-		
+  if (file.is_open())
+  {
```
Comment (olbender) [inline code/system/application/perception/src/detectlane/detectlane.cpp]: Brackets should be on the same line as the statement (exception after methods/functions, but they are not statements really)

File: code/system/application/perception/src/detectlane/detectlane.cpp [hunk 9]
```diff
@@ -70,266 +101,359 @@ void DetectLane::nextContainer(odcore::data::Container &c)
...
+  Eigen::Vector3d leftPoint2(colLeft2, row2, 1);
+  TransformPointToGlobalFrame(leftPoint2);
+
+  Eigen::Vector3d rightPoint2(colRight2, row2, 1);  
+  TransformPointToGlobalFrame(rightPoint2);
+
+  double dxPoint2 = (leftPoint2(0) + rightPoint2(0) ) / 2.0;
+  double dyPoint2 = (leftPoint2(1) + rightPoint2(1) ) / 2.0;
+
+  return std::atan2(dxPoint1-dxPoint2,dyPoint1-dyPoint2);
```
Comment (olbender) [inline code/system/application/perception/src/detectlane/detectlane.cpp:454]: Always put spaces around operators ' + ', and after comma ', '.

File: code/tools/vision/include/projection/InversePerspectiveMapping.hpp [hunk 10]
```diff
@@ -0,0 +1,48 @@
+#ifndef __INVERSEPERSPECTIVEMAPPING_HPP_
```
Comment (olbender) [inline code/tools/vision/include/projection/InversePerspectiveMapping.hpp]: Here it should be PROJECTION_INVERSEPERSPECTIVEMAPPING_HPP_

File: code/tools/vision/include/projection/projection.hpp [hunk 11]
```diff
@@ -29,7 +29,7 @@
 #include "opencv2/video/tracking.hpp"
 #include "opencv2/imgproc/imgproc.hpp"
 #include "opencv2/highgui/highgui.hpp"
-
+#include "projection/InversePerspectiveMapping.hpp"
```
Comment (olbender) [inline code/tools/vision/include/projection/projection.hpp:32]: Always put vertical space around different types of content, e.g. between include statements and namespace declarations

File: code/system/application/knowledge/include/linguistics/v2vcam/v2vcam.hpp (hunk 1 above)
Comment (bjornborg) [inline code/system/application/knowledge/include/linguistics/v2vcam/v2vcam.hpp]: Replaced with m_timeType2004.

File: code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp (hunk 2 above)
Comment (bjornborg) [inline code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp]: Fixed.

File: code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp (hunk 3 above)
Comment (bjornborg) [inline code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp]: Fixed.

File: code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp (hunk 4 above)
Comment (bjornborg) [inline code/system/application/perception/include/detectlane/InversePerspectiveMapping.hpp]: Fixed.

File: code/system/application/perception/src/detectlane/InversePerspectiveMapping.cpp (hunk 6 above)
Comment (bjornborg) [inline code/system/application/perception/src/detectlane/InversePerspectiveMapping.cpp]: Fixed.

File: code/system/application/perception/include/detectlane/detectlane.hpp (hunk 5 above)
Comment (bjornborg) [inline code/system/application/perception/include/detectlane/detectlane.hpp:57]: Fixed.

File: code/system/application/perception/src/detectlane/InversePerspectiveMapping.cpp (hunk 7 above)
Comment (bjornborg) [inline code/system/application/perception/src/detectlane/InversePerspectiveMapping.cpp]: Should be fixed. (Hard to find all files, check again)

File: code/system/application/perception/src/detectlane/detectlane.cpp (hunk 8 above)
Comment (bjornborg) [inline code/system/application/perception/src/detectlane/detectlane.cpp]: Fixed.

File: code/system/application/perception/src/detectlane/detectlane.cpp (hunk 9 above)
Comment (bjornborg) [inline code/system/application/perception/src/detectlane/detectlane.cpp:454]: Most of the code is fixed, but not all files.

File: code/tools/vision/include/projection/InversePerspectiveMapping.hpp (hunk 10 above)
Comment (bjornborg) [inline code/tools/vision/include/projection/InversePerspectiveMapping.hpp]: Fixed.

File: code/tools/vision/include/projection/projection.hpp (hunk 11 above)
Comment (bjornborg) [inline code/tools/vision/include/projection/projection.hpp:32]: Should be fixed

**Pull request merged**